Dockerfile.build

```diff
@@ -1,74 +1,87 @@
+RUN yum install -y https://repo.ius.io/ius-release-el7.rpm centos-release-scl \
+    && yum update -y \
+    && yum install -y devtoolset-10-gcc devtoolset-10-gcc-c++ git236
+ENV PATH /opt/rh/devtoolset-10/root/usr/bin:$PATH
 
-# install go
-ADD https://dl.google.com/go/go${GOLANG_VERSION}.linux-$TARGETARCH.tar.gz /tmp/go${GOLANG_VERSION}.tar.gz
-RUN mkdir -p /usr/local && tar xz -C /usr/local </tmp/go${GOLANG_VERSION}.tar.gz
+ADD https://github.com/Kitware/CMake/releases/download/v$CMAKE_VERSION/cmake-$CMAKE_VERSION-linux-x86_64.tar.gz /tmp/cmake-$CMAKE_VERSION.tar.gz
+RUN tar -zx -C /usr --strip-components 1 </tmp/cmake-$CMAKE_VERSION.tar.gz
+
+ADD https://dl.google.com/go/go$GO_VERSION.linux-$TARGETARCH.tar.gz /tmp/go-$GO_VERSION.tar.gz
+RUN mkdir -p /usr/local && tar xz -C /usr/local </tmp/go-$GO_VERSION.tar.gz
 
-# build the final binary
 WORKDIR /go/src/github.com/jmorganca/ollama
 COPY . .
+RUN /usr/local/go/bin/go generate ./...
+
+FROM --platform=linux/arm64 nvidia/cuda:$CUDA_VERSION-devel-rockylinux8 AS cuda-build-arm64
+
+ARG TARGETARCH
+ARG GO_VERSION
+ARG CMAKE_VERSION
+
+RUN yum install -y git cmake
+
+ADD https://dl.google.com/go/go$GO_VERSION.linux-$TARGETARCH.tar.gz /tmp/go-$GO_VERSION.tar.gz
+RUN mkdir -p /usr/local && tar xz -C /usr/local </tmp/go-$GO_VERSION.tar.gz
+
+WORKDIR /go/src/github.com/jmorganca/ollama
+COPY . .
+RUN /usr/local/go/bin/go generate ./...
+
+FROM --platform=linux/amd64 rocm/dev-centos-7:$ROCM_VERSION-complete AS rocm-build-amd64
+
+ARG TARGETARCH
+ARG GO_VERSION
+ARG CMAKE_VERSION
+
+RUN yum install -y https://repo.ius.io/ius-release-el7.rpm centos-release-scl \
+    && yum update -y \
+    && yum remove -y git \
+    && yum install -y devtoolset-10-gcc devtoolset-10-gcc-c++ git236
+ENV PATH /opt/rh/devtoolset-10/root/usr/bin:$PATH
 
-ENV GOOS=linux
-ENV GOARCH=$TARGETARCH
-ENV GOFLAGS=$GOFLAGS
-ENV CGO_CFLAGS=${CGO_CFLAGS}
+ADD https://github.com/Kitware/CMake/releases/download/v$CMAKE_VERSION/cmake-$CMAKE_VERSION-linux-x86_64.tar.gz /tmp/cmake-$CMAKE_VERSION.tar.gz
+RUN tar -zx -C /usr --strip-components 1 </tmp/cmake-$CMAKE_VERSION.tar.gz
+
+ADD https://dl.google.com/go/go$GO_VERSION.linux-$TARGETARCH.tar.gz /tmp/go-$GO_VERSION.tar.gz
+RUN mkdir -p /usr/local && tar xz -C /usr/local </tmp/go-$GO_VERSION.tar.gz
+
+WORKDIR /go/src/github.com/jmorganca/ollama
+COPY . .
+RUN /usr/local/go/bin/go generate ./...
+
+FROM --platform=linux/amd64 golang:$GO_VERSION AS build-amd64
+
+ARG GOFLAGS
+ARG CGO_FLAGS
+
+WORKDIR /go/src/github.com/jmorganca/ollama
+COPY . .
+COPY --from=cuda-build-amd64 /go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/cpu/lib llm/llama.cpp/build/linux/cpu/lib
+COPY --from=cuda-build-amd64 /go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/cuda/lib llm/llama.cpp/build/linux/cuda/lib
+COPY --from=rocm-build-amd64 /go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/rocm/lib llm/llama.cpp/build/linux/rocm/lib
```

Comment:
This might wind up being a bit fragile as we refine the variants, but I'm not sure if there's a good wildcard model here that doesn't make it messier.

Reply:
I agree it's not ideal. I'm open to suggestions

Reply:
I'm working on a change to enable variants support but it will probably take another day or so to get it working fully.  Lets merge this as is and I'll play around a bit to see if I can find a generalized pattern here to pick them up dynamically.

Reply:
One way is to turn each of these build stages into separate docker images, similar to how the MacOS build works

Something like this:

```
docker build -f Dockerfile.cuda -t ollama:build-cuda11 --build-arg CUDA_VERSION=11.8.0 .
docker build -f Dockerfile.rocm -t ollama:build-rocm5 --build-arg ROCM_VERSION=5.7 .
docker build -f Dockerfile.rocm -t ollama:build-rocm6 --build-arg ROCM_VERSION=6.0 .

docker create --name cuda11 ollama/build-cuda11
docker create --name rocm5 ollama/build-rocm5
docker create --name rocm6 ollama/build-rocm6

docker cp cuda11:/go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/cpu/lib llm/llama.cpp/build/linux/cpu/lib
docker cp cuda11:/go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/cuda/lib llm/llama.cpp/build/linux/cuda/lib
docker cp rocm5:/go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/rocm/lib llm/llama.cpp/build/linux/rocm5/lib
docker cp rocm6:/go/src/github.com/jmorganca/ollama/llm/llama.cpp/build/linux/rocm/lib llm/llama.cpp/build/linux/rocm6/lib
```